src/algorithm/minkowski/epa/epa2d.rs

```diff
@@ -0,0 +1,282 @@
+                &e.normal,
+            );
+            let d = p.v.dot(e.normal);
+            if d - e.distance < tolerance {
+                return Some(Contact::new_with_point(
+                    CollisionStrategy::FullResolution,
+                    e.normal,
+                    e.distance,
+                    point(&simplex, &e),
+                ));
+            } else {
+                simplex.insert(e.index, p);
+            }
+            i += 1;
+            if i >= MAX_ITERATIONS {
+                return Some(Contact::new_with_point(
+                    CollisionStrategy::FullResolution,
+                    e.normal,
+                    e.distance,
+                    point(&simplex, &e),
+                ));
+            }
+        }
+    }
+
+    fn new() -> Self {
+        Self {
+            m: marker::PhantomData,
+        }
+    }
+}
+
+/// This function returns the contact point in world space coordinates on shape A.
+///
+/// Compute the closest point to the origin on the given simplex edge, then use that to interpolate
+/// the support points coming from the A shape.
+fn point<S>(simplex: &Vec<SupportPoint<Point2<S>>>, edge: &Edge<S>) -> Point2<S>
+where
+    S: BaseFloat,
+{
+    let b = &simplex[edge.index];
+    let a = if edge.index == 0 {
+        &simplex[simplex.len() - 1]
+    } else {
+        &simplex[edge.index - 1]
+    };
+    let oa = -a.v;
+    let ab = b.v - a.v;
+    let t = oa.dot(ab) / ab.magnitude2();
+    if t < S::zero() {
+        a.sup_a.clone()
+    } else if t > S::one() {
+        b.sup_a.clone()
+    } else {
+        a.sup_a + (b.sup_a - a.sup_a) * t
+    }
+}
+
+#[derive(Debug)]
+struct Edge<S> {
+    pub normal: Vector2<S>,
+    pub distance: S,
+    pub index: usize,
+}
+
+impl<S> Edge<S>
+where
+    S: BaseFloat,
+{
+    pub fn new() -> Self {
+        Self {
+            normal: Vector2::zero(),
+            distance: S::infinity(),
+            index: 0,
+        }
+    }
+}
+
+fn closest_edge<S>(simplex: &Vec<SupportPoint<Point2<S>>>) -> Option<Edge<S>>
+where
+    S: BaseFloat,
+{
+    if simplex.len() < 3 {
+        None
+    } else {
+        let mut edge = Edge::new();
+        for i in 0..simplex.len() {
+            let j = if i + 1 == simplex.len() { 0 } else { i + 1 };
+            let a = simplex[i].v;
+            let b = simplex[j].v;
+            let e = b - a;
+            let oa = a;
+            let n = triple_product(&e, &oa, &e).normalize();
+            let d = n.dot(a);
+            if d < edge.distance {
+                edge.distance = d;
+                edge.normal = n;
+                edge.index = j;
+            }
+        }
+        Some(edge)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use cgmath::{Basis2, Decomposed, Point2, Rad, Rotation2, Vector2};
+
+    use super::*;
+    use algorithm::minkowski::SupportPoint;
+    use primitive::*;
+
+    #[test]
+    fn test_closest_edge_0() {
+        assert!(closest_edge::<f32>(&vec![]).is_none())
```

Comment:
nit: I think `assert_eq!(None, ...)` is slightly better for giving you more information on failures

Reply:
That requires PartialEq to be defined on Edge. In this case that's fine as the type parameter is BaseFloat, but for a lot of cases that's undesirable.